sympy/core/function.py

```diff
@@ -1276,12 +1277,11 @@ def __new__(cls, expr, *variables, **kwargs):
         # derivative.
         variable_count = []
         array_likes = (tuple, list, Tuple)
-        integer_likes = (int, Integer)
 
         from sympy.tensor.array import Array, NDimArray
 
         for i, v in enumerate(variables):
-            if isinstance(v, integer_likes):
+            if isinstance(v, (int, Integer, numbers.Integral)):
```

Comment:
`numbers.Integral` should be correct here, since it ensures you can perform addition and convert to `int`, which should be all we really need here. The question is, do we really want to start supporting this, if we do it somewhere, it would imply that it should be possible in other places too.

SymPy should probably be clearer on what types it allows in its API and when it becomes the responsibility of the user/downstream library to make sure the types are correct.

Reply:
It looks like sage has been using sympy like this for some time. Really what sage depends on is sympify turning its types into sympy types. Usually that's fine because we can just sympify things. The problem here is that sympify isn't being used at the outset because there was a problem with doing that hence the changes in #22551. Maybe it would still be possible to use sympify though but with more careful logic rather than just passing everything in.

Reply:
I think if we check whether `v` is an `array_like` first, then we can safely use `_sympify` on the remaining `v`. Or just try to sympify and if it fails check for `array_like`s

Reply:
> The question is, do we really want to start supporting this

It was supported; it's a regression introduced in the SymPy 1.10 series

Reply:
It was supported via sympify rather than in terms of the numbers ABCs. It previously worked because the sage Integer type defines `_sympy_` but would fail with some other class that happened to implement the `numbers.Integral` ABC.

Reply:
The support was that these arguments were converted with `sympify`, not that anything which passes an instance check with `numbers.Integral` is allowed here. I think that means that this does not only revert the regression, but also extends what types are allowed.

Reply:
Something like this should handle any class/object that is correctly set up to work with SymPy but nothing else.

```diff
diff --git a/sympy/core/function.py b/sympy/core/function.py
index 171231ce24..bc70aaba92 100644
--- a/sympy/core/function.py
+++ b/sympy/core/function.py
@@ -47,7 +47,7 @@
 from .parameters import global_parameters
 from .rules import Transform
 from .singleton import S
-from .sympify import sympify
+from .sympify import sympify, SympifyError, _sympify

 from .sorting import default_sort_key, ordered
 from sympy.utilities.exceptions import (sympy_deprecation_warning,
@@ -1276,12 +1276,16 @@ def __new__(cls, expr, *variables, **kwargs):
         # derivative.
         variable_count = []
         array_likes = (tuple, list, Tuple)
-        integer_likes = (int, Integer)

         from sympy.tensor.array import Array, NDimArray

         for i, v in enumerate(variables):
-            if isinstance(v, integer_likes):
+            try:
+                sv = _sympify(v)
+            except SympifyError:
+                sv = v
+            if isinstance(sv, Integer):
+                v = sv
                 if i == 0:
                     raise ValueError("First variable cannot be a number: %i" % v)
                 count = v
```

Reply:
Presumably the reason you didn't do it like that before was to avoid calling `sympify` on `v` if it was an array-like. I think it probably makes sense just to invert the logic in the loop so that `array_likes` is checked first and then `_sympify` is called after.

Reply:
Actually, `Basic._do_eq_sympify` could also be used here, although it seems that I have chosen the name a bit unfortunately. maybe `_do_external_sympify` would have been better.

Inverting the logic didn't seem trivial to me btw, there is a reason why it isn't just one large if-elif block.

Reply:
> The support was that these arguments were converted with `sympify`, not that anything which passes an instance check with `numbers.Integral` is allowed here. I think that means that this does not only revert the regression, but also extends what types are allowed.

It does not actually extend it: `sympify` is able to handle all instances of `numbers.Integral` because the method `__int__` is defined.
```
>>> class Eight():
...     def __int__(self): return 8
... 
>>> sympy.sympify(Eight())
8
>>> type(_)
<class 'sympy.core.numbers.Integer'>
```

Reply:
Perhaps to put it another way: it would be better if all handling of foreign types was done in one place. Ideally this function should depend on sympify doing the right thing and then it shouldn't need to worry about things like the Integral ABC. If sympify can handle the numbers ABCs then the rest of sympy can benefit from that.